qiskit/providers/__init__.py

```diff
@@ -404,6 +404,54 @@ def _define(self):
    transpiler will ensure that it continues to be well supported by Qiskit
    moving forward.
 
+.. _custom_transpiler_backend:
+
+Custom Transpiler Passes
+^^^^^^^^^^^^^^^^^^^^^^^^
+As part of the transpiler there is a provision for backends to provide custom
+stage implementation to facilitate hardware specific optimizations and
+circuit transformations. Currently there are two hook points supported,
+``get_post_translation_stage()`` which is used for a backend to specify a
+:class:`~PassManager` which will be run after basis translation stage in the
+compiler and ``get_scheduling_stage()`` which is used for a backend to
+specify a :class:`~PassManager` which will be run for the scheduling stage
+by default (which is the last defined stage in a default compilation). These
+hook points in a :class:`~.BackendV2` class should only be used if your
+backend has special requirements for compilation that are not met by the
+default backend
+
+To leverage these hook points you just need to add the methods to your
+:class:`~.BackendV2` implementation and have them return a
+:class:`~.PassManager` object. For example::
+
+    from qiskit.circuit.library import XGate
+    from qiskit.transpiler.passes import (
+        ALAPScheduleAnalysis,
+        PadDynamicalDecoupling,
+        ResetAfterMeasureSimplification
+    )
+
+    class Mybackend(BackendV2):
+
+        def get_scheduling_stage(self):
+            dd_sequence = [XGate(), XGate()]
+            pm = PassManager([
+                ALAPScheduleAnalysis(self.instruction_durations),
+                PadDynamicalDecoupling(self.instruction_durations, dd_sequence)
+            ])
+            return pm
+
+        def get_post_translation_stage(self):
+            pm = PassManager([ResetAfterMeasureSimplification()])
+            return pm
+
+This snippet of a backend implementation will now have the :func:`~.transpile`
+function run a custom stage for scheduling (unless the user manually requests a
+different one explicitly) which will insert dynamical decoupling sequences and
+also simplify resets after measurements after the basis translation stage. This
+way if these two compilation steps are **required** for running on ``Mybackend``
+the transpiler will be able to perform these steps without any manual user input.
```

Comment:
Are we planning to add a different interface for passes that are suggested but not required?  For example, `ResetAfterMeasureSimplification` is (presumably) never _required_, but for certain backends, it's a generally good idea.

Reply:
I guess required is probably too strong a word. I was trying to reinforce here that these passes will be run by default for every `transpile()` call so only use it for things you want to be run all the time. My intent was for something like `ResetAfterMeasureSimplification` to be usable in `post-translation` which is why I put it in the example. For the appropriate backend it's an easy optimization that improves the fidelity, but that is a backend specific thing and only the backend object will have sufficient context to know whether it's a good idea or not. It's not something we can cleanly express via the target (which is the point of having this interface).

But, writing this I just realize that I think we should add an optimization level kwarg to the hook method. Because, I realize to really support using `ResetAfterMeasureSimplification` we want to not inject that into the pipeline for level 0. So giving backends the context of which optimization level the pass will enable doing this and potential tweak the custom stages in a similar way for higher optimization levels.